diff --git a/bfi.cpp b/bfi.cpp
--- a/bfi.cpp
+++ b/bfi.cpp
@@ -781,11 +781,13 @@
                        IARG_CALL_ORDER, CALL_ORDER_LAST,
                        IARG_END);
         INS_RewriteMemoryOperand(ins, op, scratch);
-        insert_trigger(ins);
-        INS_InsertThenCall(ins, IPOINT_AFTER, (AFUNPTR) inject_bp,
-                           IARG_CONTEXT,
-                           IARG_THREAD_ID,
-                           IARG_END);
+        if (INS_HasFallThrough(ins)) {
+          insert_trigger(ins);
+          INS_InsertThenCall(ins, IPOINT_AFTER, (AFUNPTR) inject_bp,
+                             IARG_CONTEXT,
+                             IARG_THREAD_ID,
+                             IARG_END);
+        }
         break;
     }
     default:
