diff --git a/bfi.cpp b/bfi.cpp
--- a/bfi.cpp
+++ b/bfi.cpp
@@ -332,15 +332,12 @@
     // add an absolute jump back to the next instruction using r13
     memcpy(text + size, trampoline, sizeof(trampoline));
 
-    // adapt mask to fit instruction size
-    assert (size <= 8);
-    if (size < 8) {
-        ADDRINT tmp = mask % (1<<size);
-        // if the mask became 0, use mask 1 instead (unless it was
-        // given as 0 from user input)
-        if (tmp == 0 && mask != 0) mask = 0x01;
-        else mask = tmp;
-    }
+    // truncate mask to a single byte
+    ADDRINT tmp = mask & 0xFF;
+    // if the mask became 0, use mask 1 instead (unless it was
+    // given as 0 from user input)
+    if (tmp == 0 && mask != 0) mask = 0x01;
+    else mask = tmp;
 
     // get target byte and make idx fit instruction size
     int idx;
